sklearn/tree/_splitter.pyx

```diff
@@ -351,18 +389,28 @@ cdef class BestSplitter(BaseDenseSplitter):
                 # f_j in the interval [n_known_constants, f_i - n_found_constants[
                 f_j += n_found_constants
                 # f_j in the interval [n_total_constants, f_i[
-
                 current.feature = features[f_j]
 
-                # Sort samples along that feature; first copy the feature
-                # values for the active samples into Xf, s.t.
-                # Xf[i] == X[samples[i], j], so the sort uses the cache more
+                # Sort samples along that feature; either by utilizing
+                # presorting, or by copying the values into an array and
+                # sorting the array in a manner which utilizes the cache more
                 # effectively.
-                for p in range(start, end):
-                    Xf[p] = X[X_sample_stride * samples[p] +
-                              X_fx_stride * current.feature]
+                if self.presort == 1:
+                    p = start
+                    feature_offset = self.X_idx_sorted_stride * current.feature
+                    for i in range(self.n_total_samples): 
+                        j = X_idx_sorted[i + feature_offset]
+                        if sample_mask[j] == 1:
+                            samples[p] = j
+                            Xf[p] = X[self.X_sample_stride * j +
+                                      self.X_feature_stride * current.feature]
```

Comment:
Could you explain? Otherwise, you can do once the computation of `self.X_feature_stride * current.feature`.